File: NProcGen/src/PonderingProgrammer.Map2d/FixedSquareMap2d.cs
using System;
using System.Collections.Generic;

namespace PonderingProgrammer.Map2d
{
    public abstract class FixedSquareMap2d<T> : AbstractMap2d<T>
    {
        private readonly Bounds _bounds;
        private readonly List<Cell<T>> _cells;

        public FixedSquareMap2d(int width, int height)
        {
            _bounds = new Bounds(0, 0, width, height);

            _cells = new List<Cell<T>>(width * height);
            for (var y = 0; y < height; y++)
                for (var x = 0; x < width; x++)
                    _cells.Add(new Cell<T>(new Coord(x, y), default));
        }

        public override Bounds GetBounds() => _bounds;

        public override Cell<T> GetCell(int x, int y) => HasCell(x, y) ? _cells[CoordToIndex(x, y)] : null;

        public override IEnumerable<Cell<T>> GetCells() => _cells;

        public override IEnumerable<Cell<T>> GetCellsRowMajor() => _cells;

        public override T GetValueOrDefault(int x, int y) => HasCell(x, y) ? _cells[CoordToIndex(x, y)].Value : default;

        public override bool HasCell(int x, int y)
        {
            return x >= 0 && x < _bounds.width && y >= 0 && y < _bounds.height;
        }

        public override IEnumerable<Cell<T>> FindCells(Predicate<T> predicate) => _cells.FindAll(cell => predicate.Invoke(cell.Value));

        protected int CoordToIndex(int x, int y) => (y * _bounds.width) + x;
    }
}

File: projects/java/map2d/src/main/java/net/ponderingprogrammer/map2d/FixedSquareMap2d.java
package net.ponderingprogrammer.map2d;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Optional;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public abstract class FixedSquareMap2d<T> extends AbstractMap2d<T> {
    private final int width;
    private final int height;
    private final ArrayList<Cell<T>> cells; // flat list of cells, row-major
    private final Bounds bounds;

    public FixedSquareMap2d(int width, int height) {
        if (width < 1 || height < 1)
            throw new IllegalArgumentException(String.format("Zero or negative size is illegal (width: %d, height: %d)", width, height));

        this.width = width;
        this.height = height;

        cells = new ArrayList<>(width * height);
        for (int y = 0; y < height; y++) {  // rows
            for (int x = 0; x < width; x++) {   // cols
                cells.add(new Cell<>(new Coord(x, y), null));
            }
        }

        bounds = new Bounds(0, 0, width, height);
    }

    @Override
    public boolean hasCell(int x, int y) {
        return x >= 0 && x < width && y >= 0 && y < height;
    }

    @Override
    public Optional<Cell<T>> getCell(int x, int y) {
        return hasCell(x, y) ? Optional.of(cells.get(coordToIndex(x, y))) : Optional.empty();
    }

    @Override
    public Optional<T> getValue(int x, int y) {
        return getCell(x, y).map(Cell::getValue);
    }

    @Override
    public Collection<Cell<T>> getCells() {
        return Collections.unmodifiableList(cells);
    }

    @Override
    public Collection<Cell<T>> getCellsRowMajor() {
        return Collections.unmodifiableList(cells);
    }

    @Override
    public Optional<Bounds> getBounds() {
        return Optional.of(bounds);
    }

    @Override
    public Collection<Cell<T>> findCells(Predicate<T> predicate) {
        return cells.stream().filter(cell -> predicate.test(cell.getValue())).collect(Collectors.toCollection(ArrayList::new));
    }

    protected int coordToIndex(int x, int y) {
        return (y * width) + x;
    }
}
